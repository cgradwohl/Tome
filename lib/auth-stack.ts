import { LambdaRestApi, CfnAuthorizer, LambdaIntegration, AuthorizationType } from '@aws-cdk/aws-apigateway';
import { AssetCode, Function, Runtime } from '@aws-cdk/aws-lambda';
import { App, Stack, CfnOutput } from '@aws-cdk/core';
import { UserPool, UserPoolClient } from '@aws-cdk/aws-cognito'

export class AuthStack extends Stack {
  constructor(app: App, id: string) {
    super(app, id);

    // Cognito User Pool with Email Sign-in Type.
    const userPool = new UserPool(this, 'TomeUserPool', {
      signInAliases: {
        email: true
      }
    })

    const userPoolClient = new UserPoolClient(this, 'TomeUserPoolClient', {
      userPoolClientName: 'testClient',
      userPool: userPool,
      authFlows: {
        adminUserPassword: true,
        refreshToken: true,
      },
      preventUserExistenceErrors: true
    });

    // Function that returns 201 with "Hello world!"
    const helloWorldFunction = new Function(this, 'helloWorldFunction', {
      code: new AssetCode('lambdas'),
      handler: 'helloworld.handler',
      runtime: Runtime.NODEJS_12_X,
      // this is the only way I have been able to get the pool id
      environment: {
        USER_POOL_ID: userPool.userPoolId,
      }
    });

    const loginFunction = new Function(this, 'login', {
      code: new AssetCode('lambdas'),
      handler: 'login.handler',
      runtime: Runtime.NODEJS_12_X,
      environment: {
        USER_POOL_ID: userPool.userPoolId,
        USER_POOL_CLIENT_ID: userPoolClient.userPoolClientId, 
      }
    })

    // Rest API backed by the helloWorldFunction
    const helloWorldLambdaRestApi = new LambdaRestApi(this, 'helloWorldLambdaRestApi', {
      restApiName: 'Hello World API',
      handler: helloWorldFunction,
      proxy: false,
    });

    const loginLambdaRestApi = new LambdaRestApi(this, 'loginLambdaRestApi', {
      restApiName: 'Login API',
      handler: loginFunction,
      proxy: false,
    })

    // Authorizer for the Hello World API that uses the
    // Cognito User pool to Authorize users.
    const authorizer = new CfnAuthorizer(this, 'cfnAuth', {
      restApiId: helloWorldLambdaRestApi.restApiId,
      name: 'HelloWorldAPIAuthorizer',
      type: 'COGNITO_USER_POOLS',
      identitySource: 'method.request.header.Authorization',
      providerArns: [userPool.userPoolArn],
    })

    // Hello Resource API for the REST API. 
    const hello = helloWorldLambdaRestApi.root.addResource('HELLO');

    const login = loginLambdaRestApi.root.addResource('login');

    // GET method for the HELLO API resource. It uses Cognito for
    // authorization and the auathorizer defined above.
    hello.addMethod('GET', new LambdaIntegration(helloWorldFunction), {
      authorizationType: AuthorizationType.COGNITO,
      authorizer: {
        authorizerId: authorizer.ref
      }
      
    })

    login.addMethod('POST', new LambdaIntegration(loginFunction))

    new CfnOutput(this, "UserPoolIdOutput", {
      description: "UserPool ID",
      value: userPool.userPoolId
    });
    
  }
}


const app = new App();
new AuthStack(app, 'AuthStack');
app.synth();