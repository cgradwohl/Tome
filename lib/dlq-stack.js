"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudwatch = require("@aws-cdk/aws-cloudwatch");
const sqs = require("@aws-cdk/aws-sqs");
class DeadLetterQueue extends sqs.Queue {
    constructor(scope, id) {
        super(scope, id);
        // Add the alarm
        this.messagesInQueueAlarm = new cloudwatch.Alarm(this, 'Alarm', {
            alarmDescription: 'There are messages in the Dead Letter Queue',
            evaluationPeriods: 1,
            threshold: 1,
            metric: this.metricApproximateNumberOfMessagesVisible(),
        });
    }
}
exports.DeadLetterQueue = DeadLetterQueue;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiZGxxLXN0YWNrLmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiZGxxLXN0YWNrLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBQUEsc0RBQXNEO0FBQ3RELHdDQUF3QztBQUd4QyxNQUFhLGVBQWdCLFNBQVEsR0FBRyxDQUFDLEtBQUs7SUFHNUMsWUFBWSxLQUFnQixFQUFFLEVBQVU7UUFDdEMsS0FBSyxDQUFDLEtBQUssRUFBRSxFQUFFLENBQUMsQ0FBQztRQUVqQixnQkFBZ0I7UUFDaEIsSUFBSSxDQUFDLG9CQUFvQixHQUFHLElBQUksVUFBVSxDQUFDLEtBQUssQ0FBQyxJQUFJLEVBQUUsT0FBTyxFQUFFO1lBQzlELGdCQUFnQixFQUFFLDZDQUE2QztZQUMvRCxpQkFBaUIsRUFBRSxDQUFDO1lBQ3BCLFNBQVMsRUFBRSxDQUFDO1lBQ1osTUFBTSxFQUFFLElBQUksQ0FBQyx3Q0FBd0MsRUFBRTtTQUN4RCxDQUFDLENBQUM7SUFDTCxDQUFDO0NBQ0Y7QUFkRCwwQ0FjQyIsInNvdXJjZXNDb250ZW50IjpbImltcG9ydCAqIGFzIGNsb3Vkd2F0Y2ggZnJvbSAnQGF3cy1jZGsvYXdzLWNsb3Vkd2F0Y2gnO1xuaW1wb3J0ICogYXMgc3FzIGZyb20gJ0Bhd3MtY2RrL2F3cy1zcXMnO1xuaW1wb3J0IHsgQ29uc3RydWN0LCBEdXJhdGlvbiB9IGZyb20gJ0Bhd3MtY2RrL2NvcmUnO1xuXG5leHBvcnQgY2xhc3MgRGVhZExldHRlclF1ZXVlIGV4dGVuZHMgc3FzLlF1ZXVlIHtcbiAgcHVibGljIHJlYWRvbmx5IG1lc3NhZ2VzSW5RdWV1ZUFsYXJtOiBjbG91ZHdhdGNoLklBbGFybTtcblxuICBjb25zdHJ1Y3RvcihzY29wZTogQ29uc3RydWN0LCBpZDogc3RyaW5nKSB7XG4gICAgc3VwZXIoc2NvcGUsIGlkKTtcblxuICAgIC8vIEFkZCB0aGUgYWxhcm1cbiAgICB0aGlzLm1lc3NhZ2VzSW5RdWV1ZUFsYXJtID0gbmV3IGNsb3Vkd2F0Y2guQWxhcm0odGhpcywgJ0FsYXJtJywge1xuICAgICAgYWxhcm1EZXNjcmlwdGlvbjogJ1RoZXJlIGFyZSBtZXNzYWdlcyBpbiB0aGUgRGVhZCBMZXR0ZXIgUXVldWUnLFxuICAgICAgZXZhbHVhdGlvblBlcmlvZHM6IDEsXG4gICAgICB0aHJlc2hvbGQ6IDEsXG4gICAgICBtZXRyaWM6IHRoaXMubWV0cmljQXBwcm94aW1hdGVOdW1iZXJPZk1lc3NhZ2VzVmlzaWJsZSgpLFxuICAgIH0pO1xuICB9XG59Il19