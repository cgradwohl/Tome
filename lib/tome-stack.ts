import * as cdk from '@aws-cdk/core';
import * as cognito from '@aws-cdk/aws-cognito';
export class TomeStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here

    // Auth:
    // Cognito User Pool
    // note this is from the construct library and not a CFN Construct
    const userPool = new cognito.UserPool(this, 'TomeUserPool', {
      signInAliases: {
        email: true,
      },
      signInCaseSensitive: false,
      autoVerify: {
        email: true,
      },
      passwordPolicy: {
        minLength: 8,
        requireLowercase: true,
        requireDigits: true,
        requireUppercase: true,
        requireSymbols: true
      },
      standardAttributes: {
        givenName: {
          mutable: true,
          required: true
        },
        familyName: {
          mutable: true,
          required: true
        },
        email: {
          mutable: true,
          required: true
        }
      }
    });


    // Lambda
    // DynamoDB Table
    

    // User Data: Profile, Library, Group Meta, Queue, Tome
    // API Gateway
    // Cognito Authorizer
    // Lambda w/ read access to DDD
    // DynamoDB Table

    // TODO: folder structure (infra vs application code .... or are they the same?   :))
  }
}
