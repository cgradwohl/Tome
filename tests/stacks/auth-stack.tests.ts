import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { App, Fn } from '@aws-cdk/core';

import { AuthStack } from '../../lib/auth-stack';

import { init } from '../utils/init';

const app = new App();
const authStack = new AuthStack(app, 'TestAuthStack');



describe('auth-stack', () => {
  beforeAll(async () => {
    await init();
  })
  
  it('should work', () => {
    expect(SynthUtils.toCloudFormation(authStack)).toMatchSnapshot();
  });

  it('should deploy user pool', () => { 
    expect(authStack).toHaveResource('AWS::Cognito::UserPool', {
      AutoVerifiedAttributes: [
        "email"
      ],
      AdminCreateUserConfig: {
        AllowAdminCreateUserOnly: true,
      },
      AccountRecoverySetting: {
        RecoveryMechanisms: [
          {
            Name: "verified_phone_number",
            Priority: 1,
          },
          {
            Name: "verified_email",
            Priority: 2,
          },
        ],
      },
    });
  })
});