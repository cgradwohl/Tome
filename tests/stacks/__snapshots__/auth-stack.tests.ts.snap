// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth-stack should work 1`] = `
Object {
  "Outputs": Object {
    "UserPoolIdOutput": Object {
      "Description": "UserPool ID",
      "Value": Object {
        "Ref": "TomeUserPool39BBD8AA",
      },
    },
    "helloWorldLambdaRestApiEndpointC5601FAC": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "helloWorldLambdaRestApi6825FB98",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "helloWorldLambdaRestApiDeploymentStageprod67DD79AF",
            },
            "/",
          ],
        ],
      },
    },
    "loginLambdaRestApiEndpoint57B11974": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "loginLambdaRestApiB32E0BD2",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "loginLambdaRestApiDeploymentStageprod5BBE42A8",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900ArtifactHash27011E67": Object {
      "Description": "Artifact hash for asset \\"0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900\\"",
      "Type": "String",
    },
    "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3BucketF6815B4C": Object {
      "Description": "S3 bucket for asset \\"0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900\\"",
      "Type": "String",
    },
    "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3VersionKey19B7C235": Object {
      "Description": "S3 key for asset version \\"0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TomeUserPool39BBD8AA": Object {
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            Object {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": true,
        },
        "AutoVerifiedAttributes": Array [
          "email",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "SmsConfiguration": Object {
          "ExternalId": "TestAuthStackTomeUserPool1B88D190",
          "SnsCallerArn": Object {
            "Fn::GetAtt": Array [
              "TomeUserPoolsmsRoleC1B9FF88",
              "Arn",
            ],
          },
        },
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UsernameAttributes": Array [
          "email",
        ],
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
    },
    "TomeUserPoolClient73667605": Object {
      "Properties": Object {
        "AllowedOAuthFlows": Array [
          "implicit",
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": Array [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": Array [
          "https://example.com",
        ],
        "ClientName": "testClient",
        "ExplicitAuthFlows": Array [
          "ALLOW_ADMIN_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "PreventUserExistenceErrors": "ENABLED",
        "SupportedIdentityProviders": Array [
          "COGNITO",
        ],
        "UserPoolId": Object {
          "Ref": "TomeUserPool39BBD8AA",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "TomeUserPoolsmsRoleC1B9FF88": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Condition": Object {
                "StringEquals": Object {
                  "sts:ExternalId": "TestAuthStackTomeUserPool1B88D190",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cognito-idp.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "sns:Publish",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "sns-publish",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "cfnAuth": Object {
      "Properties": Object {
        "IdentitySource": "method.request.header.Authorization",
        "Name": "HelloWorldAPIAuthorizer",
        "ProviderARNs": Array [
          Object {
            "Fn::GetAtt": Array [
              "TomeUserPool39BBD8AA",
              "Arn",
            ],
          },
        ],
        "RestApiId": Object {
          "Ref": "helloWorldLambdaRestApi6825FB98",
        },
        "Type": "COGNITO_USER_POOLS",
      },
      "Type": "AWS::ApiGateway::Authorizer",
    },
    "helloWorldFunction00C940B5": Object {
      "DependsOn": Array [
        "helloWorldFunctionServiceRole8475DBF0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3BucketF6815B4C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3VersionKey19B7C235",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3VersionKey19B7C235",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "USER_POOL_ID": Object {
              "Ref": "TomeUserPool39BBD8AA",
            },
          },
        },
        "Handler": "helloworld.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "helloWorldFunctionServiceRole8475DBF0",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "helloWorldFunctionServiceRole8475DBF0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "helloWorldLambdaRestApi6825FB98": Object {
      "Properties": Object {
        "Name": "Hello World API",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "helloWorldLambdaRestApiAccount3D66EADA": Object {
      "DependsOn": Array [
        "helloWorldLambdaRestApi6825FB98",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "helloWorldLambdaRestApiCloudWatchRole22367FBD",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "helloWorldLambdaRestApiCloudWatchRole22367FBD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "helloWorldLambdaRestApiDeploymentCB1FF574ae10c162afd462d14276188f96f29d56": Object {
      "DependsOn": Array [
        "helloWorldLambdaRestApiHELLOGET6E88F46C",
        "helloWorldLambdaRestApiHELLOAD091B67",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "helloWorldLambdaRestApi6825FB98",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "helloWorldLambdaRestApiDeploymentStageprod67DD79AF": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "helloWorldLambdaRestApiDeploymentCB1FF574ae10c162afd462d14276188f96f29d56",
        },
        "RestApiId": Object {
          "Ref": "helloWorldLambdaRestApi6825FB98",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "helloWorldLambdaRestApiHELLOAD091B67": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "helloWorldLambdaRestApi6825FB98",
            "RootResourceId",
          ],
        },
        "PathPart": "HELLO",
        "RestApiId": Object {
          "Ref": "helloWorldLambdaRestApi6825FB98",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "helloWorldLambdaRestApiHELLOGET6E88F46C": Object {
      "Properties": Object {
        "AuthorizationType": "COGNITO_USER_POOLS",
        "AuthorizerId": Object {
          "Ref": "cfnAuth",
        },
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "helloWorldFunction00C940B5",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "helloWorldLambdaRestApiHELLOAD091B67",
        },
        "RestApiId": Object {
          "Ref": "helloWorldLambdaRestApi6825FB98",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "helloWorldLambdaRestApiHELLOGETApiPermissionTestAuthStackhelloWorldLambdaRestApiDD8D8853GETHELLOEDFC2CE9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "helloWorldFunction00C940B5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "helloWorldLambdaRestApi6825FB98",
              },
              "/",
              Object {
                "Ref": "helloWorldLambdaRestApiDeploymentStageprod67DD79AF",
              },
              "/GET/HELLO",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "helloWorldLambdaRestApiHELLOGETApiPermissionTestTestAuthStackhelloWorldLambdaRestApiDD8D8853GETHELLO6A5EDC0F": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "helloWorldFunction00C940B5",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "helloWorldLambdaRestApi6825FB98",
              },
              "/test-invoke-stage/GET/HELLO",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "login68DC89FC": Object {
      "DependsOn": Array [
        "loginServiceRole1314ABEB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3BucketF6815B4C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3VersionKey19B7C235",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0ea07ae320b2c0472513620b4b5de100b23b0cdb4bb856cf502e4c9640ac1900S3VersionKey19B7C235",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "USER_POOL_CLIENT_ID": Object {
              "Ref": "TomeUserPoolClient73667605",
            },
            "USER_POOL_ID": Object {
              "Ref": "TomeUserPool39BBD8AA",
            },
          },
        },
        "Handler": "login.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "loginServiceRole1314ABEB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "loginLambdaRestApiAccountF5DE90BD": Object {
      "DependsOn": Array [
        "loginLambdaRestApiB32E0BD2",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "loginLambdaRestApiCloudWatchRole7ADB44A5",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "loginLambdaRestApiB32E0BD2": Object {
      "Properties": Object {
        "Name": "Login API",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "loginLambdaRestApiCloudWatchRole7ADB44A5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "loginLambdaRestApiDeploymentCEEB162Dd03ae6a224e1ac39435d95b3c0837670": Object {
      "DependsOn": Array [
        "loginLambdaRestApiloginPOST3C8E5352",
        "loginLambdaRestApilogin008C6330",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "loginLambdaRestApiB32E0BD2",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "loginLambdaRestApiDeploymentStageprod5BBE42A8": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "loginLambdaRestApiDeploymentCEEB162Dd03ae6a224e1ac39435d95b3c0837670",
        },
        "RestApiId": Object {
          "Ref": "loginLambdaRestApiB32E0BD2",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "loginLambdaRestApilogin008C6330": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "loginLambdaRestApiB32E0BD2",
            "RootResourceId",
          ],
        },
        "PathPart": "login",
        "RestApiId": Object {
          "Ref": "loginLambdaRestApiB32E0BD2",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "loginLambdaRestApiloginPOST3C8E5352": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "login68DC89FC",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "loginLambdaRestApilogin008C6330",
        },
        "RestApiId": Object {
          "Ref": "loginLambdaRestApiB32E0BD2",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "loginLambdaRestApiloginPOSTApiPermissionTestAuthStackloginLambdaRestApiEE90B234POSTloginD44C8E24": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "login68DC89FC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "loginLambdaRestApiB32E0BD2",
              },
              "/",
              Object {
                "Ref": "loginLambdaRestApiDeploymentStageprod5BBE42A8",
              },
              "/POST/login",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "loginLambdaRestApiloginPOSTApiPermissionTestTestAuthStackloginLambdaRestApiEE90B234POSTloginCA9B48E0": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "login68DC89FC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "loginLambdaRestApiB32E0BD2",
              },
              "/test-invoke-stage/POST/login",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "loginServiceRole1314ABEB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
